### PR TITLE
Update Android build environment to 29 version

### PR DESCRIPTION
**Pull request by rdub-bba**

### CODE REVIEW

File: generate-sdk-packages-list.sh [hunk 1]
```diff
@@ -3,7 +3,7 @@
 SRC_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
```
Comment (alexey-pelykh) [inline generate-sdk-packages-list.sh]: Do we still need this file?

File: sdk-packages.list [hunk 2]
```diff
@@ -1,50 +1,36 @@
+cmake;3.10.2.4988404
 cmake;3.6.4111459
```
Comment (alexey-pelykh) [inline sdk-packages.list]: Do we still need this file?

File: Dockerfile [hunk 3]
```diff
@@ -1,47 +1,67 @@
...
+# Installing SDKs and other components
+# Accepting all non-standard licenses
+RUN yes | sdkmanager \
+    "cmake;3.10.2.4988404" \
+    "cmake;3.6.4111459" \
+    "platforms;android-29" \
+    "build-tools;29.0.1" \
+    "build-tools;29.0.0" \
+    "system-images;android-29;google_apis;x86" \
+    "system-images;android-28;google_apis;x86" \
```
Comment (alexey-pelykh) [inline Dockerfile]: Why do you need emulator and system images of different versions?

File: Dockerfile [hunk 4]
```diff
@@ -1,47 +1,67 @@
...
+    "system-images;android-25;google_apis;armeabi-v7a" \
+    "system-images;android-24;default;armeabi-v7a" \
+    "system-images;android-22;default;armeabi-v7a" \
+    "system-images;android-19;default;armeabi-v7a" \
+    "extras;android;m2repository" \
+    "extras;google;m2repository" \
+    "extras;google;google_play_services" \
+    "extras;m2repository;com;android;support;constraint;constraint-layout;1.0.2" \
+    "extras;m2repository;com;android;support;constraint;constraint-layout;1.0.1" \
+    "add-ons;addon-google_apis-google-23" \
```
Comment (alexey-pelykh) [inline Dockerfile]: Why do you include Google APIs 23, 22, 21 in image of platform api 29?

File: generate-sdk-packages-list.sh (hunk 1 above)
Comment (rdub-bba) [reply generate-sdk-packages-list.sh]: No, we don't. The script is going to be removed in the next commit.

Comment (alexey-pelykh) [reply generate-sdk-packages-list.sh]: Looking forward to

File: sdk-packages.list (hunk 2 above)
Comment (alexey-pelykh) [reply sdk-packages.list]: 1) I meant entirely `sdk-packages.list`
2) Most up-to-date version of cmake is good enough

Comment (rdub-bba) [reply sdk-packages.list]: 1. Ok. The sdk-packages.list and generate-sdk-packages-list.sh files are not needed anymore.
2. Will keep only the latest version of cmake.

File: Dockerfile (hunk 3 above)
Comment (rdub-bba) [reply Dockerfile]: Added by a mistake. Will be removed.

File: Dockerfile (hunk 4 above)
Comment (rdub-bba) [reply Dockerfile]: Added by a mistake. Will be removed.

File: Dockerfile [hunk 5]
```diff
@@ -1,47 +1,56 @@
...
 
-# Install OpenJDK
-# Ref: https://www.digitalocean.com/community/tutorials/how-to-install-java-with-apt-get-on-ubuntu-16-04
-RUN apt-get update \
-  && apt-get install -y --no-install-recommends openjdk-8-jdk \
-  && apt-get clean \
-  && rm -rf /var/lib/apt/lists/* \
-  && rm -rf /var/cache/oracle-jdk8-installer;
+# Updating the system
+RUN apt-get update -qq
```
Comment (alexey-pelykh) [inline Dockerfile]: Please clean the cache afterwards to reduce image size

File: Dockerfile [hunk 6]
```diff
@@ -1,47 +1,56 @@
...
 
-# Setup JAVA_HOME
+# Installing OpenJDK and x86 dependencies
+RUN dpkg --add-architecture i386
+RUN apt-get update -qq
+RUN DEBIAN_FRONTEND=noninteractive apt-get install -y openjdk-8-jdk libc6:i386 libstdc++6:i386 libgcc1:i386 libncurses5:i386 libz1:i386 \
```
Comment (alexey-pelykh) [inline Dockerfile]: Why do you install same libraries twice?

File: Dockerfile [hunk 7]
```diff
@@ -1,47 +1,56 @@
...
+RUN wget -q https://dl.google.com/android/repository/sdk-tools-linux-4333796.zip -O android-sdk-tools.zip \
+    && unzip -q android-sdk-tools.zip -d ${ANDROID_HOME} \
+    && rm android-sdk-tools.zip
+ENV PATH ${PATH}:${ANDROID_HOME}/tools:${ANDROID_HOME}/tools/bin:${ANDROID_HOME}/platform-tools
+
+# Accepting license
+RUN yes | sdkmanager --licenses
+
+# Installing platform tools
+RUN sdkmanager "emulator" "tools" "platform-tools"
```
Comment (alexey-pelykh) [inline Dockerfile]: Once again, why we need emulator?

File: Dockerfile [hunk 8]
```diff
@@ -1,47 +1,56 @@
...
-  && apt-get install -y --no-install-recommends openjdk-8-jdk \
-  && apt-get clean \
-  && rm -rf /var/lib/apt/lists/* \
-  && rm -rf /var/cache/oracle-jdk8-installer;
+# Updating the system
+RUN apt-get update -qq
 
-# Setup JAVA_HOME
+# Installing OpenJDK and x86 dependencies
+RUN dpkg --add-architecture i386
```
Comment (alexey-pelykh) [inline Dockerfile]: @darhon why do we need this?

File: Dockerfile [hunk 9]
```diff
@@ -1,47 +1,56 @@
...
+RUN apt-get update -qq
+RUN DEBIAN_FRONTEND=noninteractive apt-get install -y openjdk-8-jdk libc6:i386 libstdc++6:i386 libgcc1:i386 libncurses5:i386 libz1:i386 \
+    && rm -rf /var/cache/oracle-jdk8-installer \
+    && apt-get clean \
+    && rm -rf /var/lib/apt/lists/*
 ENV JAVA_HOME /usr/lib/jvm/java-8-openjdk-amd64/
 
 # Install Android Source dependencies
-RUN apt-get update \
```
Comment (alexey-pelykh) [inline Dockerfile]: @darhon I'm not sure if `apt-get update` on every RUN install is needed or not, it's something to caching layers but I'm not sure what's the correct way

File: Dockerfile (hunk 8 above)
Comment (rdub-bba) [reply Dockerfile]: Removed i386 packages.

File: Dockerfile (hunk 7 above)
Comment (rdub-bba) [reply Dockerfile]: Emulator is removed now

File: Dockerfile (hunk 6 above)
Comment (rdub-bba) [reply Dockerfile]: Fixed.

File: Dockerfile (hunk 5 above)
Comment (rdub-bba) [reply Dockerfile]: Fixed.

File: Dockerfile (hunk 9 above)
Comment (rdub-bba) [reply Dockerfile]: https://hackernoon.com/tips-to-reduce-docker-image-sizes-876095da3b34 - "Tip #5 Add rm -rf /var/lib/apt/lists/* to same layer as apt-get installs" which means that apt cache clean should be executed after each apt-get install command

**Comment by rdub-bba** [comment]
Updated.

**Comment by rdub-bba** [comment]
Updated.

**Comment by strikoza** [comment]
Updated.

File: Dockerfile (hunk 9 above)
Comment (alexey-pelykh) [reply Dockerfile]: @rdub-bba that's right, and that means that prior to every call to `apt-get` we need to update the cache by calling update, right? So this line needs to remain there, right?

File: Dockerfile [hunk 10]
```diff
@@ -2,63 +2,72 @@ FROM brainbeanapps/base-linux-build-environment:latest
 
 LABEL maintainer="devops@brainbeanapps.com"
 
-# Switch to root
-USER root
+ENV ANDROID_HOME /opt/android-sdk
+ENV DEBIAN_FRONTEND noninteractive
```
Comment (alexey-pelykh) [inline Dockerfile]: Isn't this line defined in parent linux environment?

File: Dockerfile [hunk 11]
```diff
@@ -2,63 +2,72 @@ FROM brainbeanapps/base-linux-build-environment:latest
...
-# Setup JAVA_HOME
+    && apt-get install --no-install-recommends -y openjdk-8-jdk \
+    && rm -rf /var/cache/oracle-jdk8-installer \
+    && apt-get clean \
+    && rm -rf /var/lib/apt/lists/*
 ENV JAVA_HOME /usr/lib/jvm/java-8-openjdk-amd64/
 
 # Install Android Source dependencies
 RUN apt-get update \
-  && apt-get install -y --no-install-recommends git-core gnupg flex bison gperf build-essential zip curl zlib1g-dev gcc-multilib g++-multilib libc6-dev-i386 lib32ncurses5-dev x11proto-core-dev libx11-dev lib32z-dev ccache libgl1-mesa-dev libxml2-utils xsltproc unzip \
```
Comment (alexey-pelykh) [inline Dockerfile]: `libc6-dev-i386` vs `libc6-dev` please elaborate? https://source.android.com/setup/build/initializing

File: Dockerfile [hunk 12]
```diff
@@ -2,63 +2,72 @@ FROM brainbeanapps/base-linux-build-environment:latest
...
 RUN curl -sL https://deb.nodesource.com/setup_8.x | bash - \
   && apt-get update \
-  && apt-get install -y --no-install-recommends nodejs \
+  && apt-get install --no-install-recommends -y nodejs \
   && apt-get clean \
   && rm -rf /var/lib/apt/lists/* \
   && npm install -g npm@latest
 
-# Install Yarn
+# Install Yarn & Ruby
```
Comment (alexey-pelykh) [inline Dockerfile]: Please install via separate chunks of commands

File: Dockerfile [hunk 13]
```diff
@@ -2,63 +2,72 @@ FROM brainbeanapps/base-linux-build-environment:latest
...
   && apt-get update \
-  && apt-get install -y --no-install-recommends yarn \
+  && apt-get install --no-install-recommends -y yarn ruby ruby-dev \
   && apt-get clean \
   && rm -rf /var/lib/apt/lists/*
 
 # Switch to user
 USER user
-WORKDIR /home/user
+WORKDIR /home/user
```
Comment (alexey-pelykh) [inline Dockerfile]: Newline, dude!

File: Dockerfile (hunk 10 above)
Comment (rdub-bba) [reply Dockerfile]: No, it doesn't.

File: Dockerfile (hunk 11 above)
Comment (rdub-bba) [reply Dockerfile]: Will 32-bit applications be building in future?

File: Dockerfile (hunk 13 above)
Comment (rdub-bba) [reply Dockerfile]: Done.

File: Dockerfile (hunk 11 above)
Comment (alexey-pelykh) [reply Dockerfile]: It's not related to building apps at all, check the referenced doc

File: Dockerfile (hunk 10 above)
Comment (alexey-pelykh) [reply Dockerfile]: https://github.com/brainbeanapps/base-linux-build-environment/blob/e529f86466e5164773967c3242807dbfd7295de8/Dockerfile#L12

https://vsupalov.com/docker-arg-vs-env/

we need it during building or running container?

Comment (rdub-bba) [reply Dockerfile]: We need DEBIAN_FRONTEND only during building. So let set it with ARG.

Comment (alexey-pelykh) [reply Dockerfile]: And ARG from parent image is not propagated?

File: Dockerfile (hunk 11 above)
Comment (rdub-bba) [reply Dockerfile]: The referenced doc is for Trusty. We used Bionic in parent image witch doesn't have i386 support from the box anymore.
I believe we should use 64-bit dependencies.

`libc6-dev-i386` used for 32-bit development libraries for AMD64 https://packages.ubuntu.com/bionic/libc6-dev-i386

It successfully work with `libc6-dev-i386` or `libc6-dev`. So we could use one of them.

Comment (alexey-pelykh) [reply Dockerfile]: Do we have a proof that this change is needed? `libc6-dev-i386` is not available anymore in Bionic?

Comment (rdub-bba) [reply Dockerfile]: This change in not needed. We may leave it as it was. `libc6-dev-i386` is available in Bionic.

File: Dockerfile (hunk 10 above)
Comment (rdub-bba) [reply Dockerfile]: No, it doesn't inherited.
We should set ENV in parent Dockerfile or use ARG or ENV in this image.

**Comment by rdub-bba** [comment]
Updated.

**Pull request closed without merging**